### ISSUE TITLE
Closing resources upon Cancelable.cancel

### ISSUE THREAD

**Issue by q42mark** [monifu/monifu#38]
Hi,

Currently using Monifu with ScalaJS and no performance issues :) Great lib! 

I am wrapping a HTML5 api called `EventSource` which is a continious stream of data. The methods;

```scala
class EventSource(URL : scala.Predef.String, settings : scala.scalajs.js.Dynamic = { /* compiled code */ }) extends org.scalajs.dom.raw.EventTarget {
  def url : scala.Predef.String = { /* compiled code */ }
  def withCredentials : scala.Boolean = { /* compiled code */ }
  def readyState : scala.Int = { /* compiled code */ }
  var onopen : scala.scalajs.js.Function1[org.scalajs.dom.raw.Event, _] = { /* compiled code */ }
  var onmessage : scala.scalajs.js.Function1[org.scalajs.dom.raw.MessageEvent, _] = { /* compiled code */ }
  var onerror : scala.scalajs.js.Function1[org.scalajs.dom.raw.Event, _] = { /* compiled code */ }
  def close() : scala.Unit = { /* compiled code */ }
}
```

When a subscription is complete (in other words, the consumer calls `Cancelable.cancel()` __BEFORE__ any `onNext` is called. How do I get a propagation of `.cancel()` inside my `Observable.create`. I need to close the EventSource connection to avoid issues :smile: 

Hope to hear from you soon

**Comment by alexandru** [comment]
@q42mark sorry for not replying, I've been traveling this week. A usage example (can be in Javascript btw), would help me advise you in the best way to wrap this. Thing is, I'm not sure what "consumer" you've got.

An important thing to note is that the design of Monifu differs from the design of RxJava / Rx.NET in the way that streaming is canceled. In RxJava you have to work with `Subscription.cancel()` to interrupt the streaming, but this `cancel()` is not available for the `Observer` unless you capture it in some way, which in my opinion is not a good design, because it's either the data source (the observable), or the consumer (the observer) that usually wants to cancel the streaming. 

In Monifu on the other hand, the streaming gets interrupted either by the data-source (the `Observable` that sends `onComplete`), or the consumer(s) (the `Observer` instances that can return a `Cancel`). In other words the interruption is not driven by logic that sits in some `Subscription` or `Cancelable` and from the point of view of the `Observer`, the cancelation is very precise - like once you return a `Cancel`, you're guaranteed to not receive any other events. This makes things a lot simpler and easier to reason about, but it means some changes.

`Observable.subscribe()` still returns a `Cancelable`. The way it is implemented though is by `Observable.takeWhileNotCanceled`. This means that this interruption may not be immediate and the source has to send an event before being interrupted. Take a look at [the source](https://github.com/monifu/monifu/blob/towards-1.0/monifu/shared/src/main/scala/monifu/reactive/operators/take.scala#L228) of `takeWhileNotCanceled` 
to see what I mean.

Now, here's the thing. The data-source you're talking about needs buffering (see [asyncBoundary](https://github.com/monifu/monifu/blob/towards-1.0/monifu/shared/src/main/scala/monifu/reactive/Observable.scala#L838)), because you cannot signal the demand from consumers in that `onmessage` (doing back-pressure in other words).

And if you need that buffer, might as well use a "*hot observable*". A hot observable is one that shares the datasource with all consumers. Subjects are hot observables. And Channels are basically subjects that also have a buffer attached. So you can use a Channel to model this.

Not sure how what the API is, but let me draw out some pseudo code ...

```scala
def buildObservable[T](eventSource: Something[T])
    (implicit s: Scheduler): (Cancelable, Observable[T]) = {

  new ConnectableObservable[T] {
    private val channel = PublishChannel[T]()

    // building the subscription
    // on first access in connect()
    private lazy val subscription = {
      // making this up, replace with whatever API you have
      val s = eventSource.handle { 
        onMessage = message => channel.pushNext(message),
        onError = ex => channel.pushError(ex)
        onComplete = channel.pushComplete()
      }
      
      BooleanCancelable {
        subscription.unsubscribe()
        channel.pushComplete()
      }
    }

    // note that connect is idempotent
    def connect(): Cancelable = {
      subscription
    }
  }
}

// ..........

val observable = buildObservable(eventSource)

// connecting consumers
observable.subscribe(observer1)
observable.subscribe(observer2)
observable.subscribe(observer3)

// starting streaming
val cancelable = observable.connect()

// stopping everything
cancelable.cancel()
```

What this does is to build a hot observable that will stream the same events to multiple subscribers.
And if you wanted, you could `.connect()` sooner and capture that reference inside `Observer` instances that you're passing to `.subscribe()`. But note that also means the connection will activate before the subscriptions and so you can lose events.

BTW - another version of Monifu is close to being released and has some API changes. In the above example the implicit scheduler will be taken by `connect()`. Mentioning this as a warning that I'll break this particular sample.

Thanks for using Monifu :-)

**Comment by q42mark** [comment]
Thank you for this extensive answer! Nice pointer on the backpressure of onMessage and Channel's. I hadn't discovered that yet. I'll code up a new event source later today and keep you posted :smile:

**Comment by q42mark** [comment]
This seems to work nicely :) 

```scala
def eventSource[T : upickle.Reader](url: String)
                                     (implicit scheduler: Scheduler): ConnectableObservable[T] = {
	new ConnectableObservable[T] {
	  private val channel = PublishChannel[T]()

	  lazy val sub = {

	    val es = new EventSource(url)

	    es.onerror = (ev: Event) => channel.pushError(new Exception("Event source got an error"))
	    es.onmessage = (ev: MessageEvent) => ev.data
	      .toString
	      .split('\n')
	      .foreach(s => channel.pushNext(upickle.read(s)))

	    BooleanCancelable {
	      es.close()
	      channel.pushComplete()
	    }
	  }

	  override def connect() = sub
	  override protected def subscribeFn(observer: Observer[T]): Unit = {
	    channel.subscribe(observer)
	  }
	} 
}
```

**Comment by alexandru** [comment]
On `|@|` I actually want to integrate Scalaz with Monifu - in other words implement some type-classes in it. This has been the primary reason for why API changes are happening in the `towards-1.0` branch - because Scala'z `flatMap` does not take an implicit execution context and this got me thinking about design :-) 

However for Javascript it is problematic to place that in the main project, as Scalaz does not support Scala.js officially yet, plus I'm a little concerned about the size of the compiled blob, as Scalaz is big.

Let me know about your needs, because the primary purpose right now are my needs :-) and some feedback would be appreciated.

On a final note - you don't really need a buffer in case you can afford to lose events. For example when tracking mouse movements, you don't necessarily need every pixel that ever happened. So you can drop events, by means of say `sample`, or `sampleRepeated` or `whileBusyDropEvents`.

**Comment by alexandru** [comment]
@q42mark - I released version `1.0-M1` to Maven Central. This is the one I was telling you it will break that sample a little. You can checkout the release notes, if you're curious, at - https://github.com/monifu/monifu/wiki/1.0-M1

Will close this issue for now - have to clean them up :-)

BTW, you can also hop on the Gitter chat, as even if I'm not online, I get notifications and will reply eventually :-)

https://gitter.im/monifu/monifu
